### ISSUE TITLE
No atomic instructions on i386

### ISSUE THREAD

**Issue by jwt27** [openssl/openssl#19323]
Atomic instructions are used in `include/internal/refcount.h` which are only
available on i486 or later.  This is an issue mainly on djgpp where i386
compatibility is generally desired.

There is a macro to disable it, but the commit that introduced it specifically
warns:

> Configure with -DOPENSSL_DEV_NO_ATOMICS and you get refcount without
> atomics.  This is intended for internal development only, to check the
> refcounting is properly coded.  It should never become a configuration
> option, hence the name of the macro.

(503d4745a115b82db01c1fb22baaddb153d27cdb)

But is this also the case when the target has no native thread support?  A user
could of course call libssl functions from interrupt context, but that should
probably not be recommended.

Is it reasonable to introduce something like the following?

```patch
--- a/include/internal/refcount.h
+++ b/include/internal/refcount.h
@@ -13,7 +13,7 @@
 # include <openssl/e_os2.h>
 # include <openssl/trace.h>

-# ifndef OPENSSL_DEV_NO_ATOMICS
+# if defined(OPENSSL_THREADS) && !defined(OPENSSL_DEV_NO_ATOMICS)
 #  if defined(__STDC_VERSION__) && __STDC_VERSION__ >= 201112L \
       && !defined(__STDC_NO_ATOMICS__)
 #   include <stdatomic.h>
```

**Comment by paulidale** [comment]
Using that define looks like it would work properly.
Configuring with `-DOPENSSL_DEV_NO_ATOMICS` ought to be enough.

I'm not sure what your suggested change brings.

**Comment by jwt27** [comment]
The difference is in intent.  There shouldn't be a `no-atomics` option and `-DOPENSSL_DEV_NO_ATOMICS` is specifically meant for development purposes.  But if a user configures with `no-threads`, that means it's safe to implicitly disable atomics, too.

**Comment by nhorman** [comment]
marking as inactive, will be closed at the end of the 3.4 dev cycle barring further input

**Comment by jwt27** [comment]
Already fixed by #19751, I forgot to link this issue # there.
